src/bsp/platform/iotc_bsp_io_net_test.cc

```diff
@@ -0,0 +1,128 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "iotc_test_echoserver.h"
+#include "gtest.h"
+#include "iotc_bsp_io_net.h"
+
+#include <stdio.h>
+
+namespace iotctest {
+namespace {
+
+typedef struct NetworkTestCase {
+  iotc_bsp_socket_type_t socket_type;
+  iotc_bsp_protocol_type_t protocol_type;
+  char* listening_addr;
+} NetworkType;
+
+class ServerTest : public ::testing::TestWithParam<NetworkType*> {
+  public:
+    const uint16_t kTestPort = 2000;
+    const char* kTestMsg = "hello\n";
+    bool WaitUntilSocketReadyForWrite();
+    bool WaitUntilSocketReadyForRead();
+    int out_written_count_, state_;
+    const char* listening_addr_;
+    iotc_bsp_socket_type_t socket_type_;
+    iotc_bsp_protocol_type_t protocol_type_;
+    iotc_bsp_socket_events_t socket_evts_[1];
+    iotc_bsp_socket_t test_socket_;
+};
+
+bool ServerTest::WaitUntilSocketReadyForWrite() {
+  bool ready_to_write = false;
+  memset(socket_evts_, 0, sizeof(iotc_bsp_socket_events_t));
+  socket_evts_[0].iotc_socket = test_socket_;
+  socket_evts_[0].in_socket_want_write = 1;
+  while (true) {
+    state_ = iotc_bsp_io_net_select(socket_evts_, 1, kTimeoutSeconds);
+    if (state_ == IOTC_BSP_IO_NET_STATE_OK &&
+        socket_evts_[0].out_socket_can_write == 1) {
+      ready_to_write = true;
+      break;
+    }
+  }
+  return ready_to_write;
+}
+
+bool ServerTest::WaitUntilSocketReadyForRead() {
+  bool ready_to_read = false;
+  memset(socket_evts_, 0, sizeof(iotc_bsp_socket_events_t));
+  socket_evts_[0].iotc_socket = test_socket_;
+  socket_evts_[0].in_socket_want_read = 1;
+  while (true) {
+    state_ = iotc_bsp_io_net_select(socket_evts_, 1, kTimeoutSeconds);
+    if (state_ == IOTC_BSP_IO_NET_STATE_OK &&
+        socket_evts_[0].out_socket_can_read == 1) {
+      ready_to_read = true;
+      break;
+    }
+  }
+  return ready_to_read;
+}
+
+TEST_P(ServerTest, EndToEndCommunicationWorks) {
+  const NetworkType* test_case = GetParam();
+  socket_type_ = test_case->socket_type;
+  listening_addr_ = test_case->listening_addr;
+  protocol_type_ = test_case->protocol_type;
+
+  std::unique_ptr<EchoTestServer> test_server(
+      new EchoTestServer(socket_type_, kTestPort, protocol_type_));
+
+  ASSERT_EQ(iotc_bsp_io_net_socket_connect(&test_socket_, listening_addr_,
+                                           kTestPort, socket_type_),
+            IOTC_BSP_IO_NET_STATE_OK);
+  ASSERT_EQ(iotc_bsp_io_net_connection_check(test_socket_, listening_addr_,
+                                             kTestPort),
+            IOTC_BSP_IO_NET_STATE_OK);
+
+  test_server->Run();
+  ASSERT_TRUE(WaitUntilSocketReadyForWrite());
+  EXPECT_EQ(iotc_bsp_io_net_write(test_socket_, &out_written_count_,
+                                  (uint8_t*)kTestMsg, strlen(kTestMsg)),
+            IOTC_BSP_IO_NET_STATE_OK);
+  char recv_buf_[strlen(kTestMsg)];
+  recv_buf_[strlen(kTestMsg)] = {0};
```

Comment:
Hm, this sets the `char` after the last element.
I ment a one-liner:
`char recv_buf_[strlen(kTestMsg)] = {0};` 
https://en.cppreference.com/w/c/language/array_initialization